### PR TITLE
Integrate Robotframework without Renode into twister

### PR DESCRIPTION
**Pull request by MP-StefanKraus**
This change adds a robot framework integration into twister, see Issue https://github.com/zephyrproject-rtos/zephyr/issues/64825

### CODE REVIEW

**Comment by nashif** [comment]
Please look at CI results and the failing tests which seem related to this change.

File: scripts/pylib/pytest-twister-harness/src/twister_harness/robot_framework/ZephyrLibrary.py
Comment (gchwier) [inline scripts/pylib/pytest-twister-harness/src/twister_harness/robot_framework/ZephyrLibrary.py:1]: It should not be a part of `pytest-twister-harness` package. It does nothing to do with pytest at all. To run internal tests for that plugin, also `robot` package must be installed now.

File: scripts/pylib/pytest-twister-harness/src/twister_harness/argument_handler.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            metavar="PATH",
+            help="Script executed after flashing.",
+        )
+        self.argument_add_func(
+            "--post-script",
+            metavar="PATH",
+            help="Script executed after closing serial connection.",
+        )
+        self.argument_add_func(
+            "--dut-scope",
```
Comment (gchwier) [inline scripts/pylib/pytest-twister-harness/src/twister_harness/argument_handler.py:83]: this parameter is rather only for pytest (uses by fixtures)

File: scripts/pylib/twister/twisterlib/harness.py [hunk 2]
```diff
@@ -284,60 +286,36 @@ def handle(self, line):
...
-        command.extend([os.path.normpath(os.path.join(
-            self.source_dir, os.path.expanduser(os.path.expandvars(src)))) for src in pytest_root])
+        self.report_file = os.path.join(self.running_dir, report_filename)
+        self.log_file_path = os.path.join(self.running_dir, log_filename)
 
-        if pytest_dut_scope:
-            command.append(f'--dut-scope={pytest_dut_scope}')
+    @abc.abstractmethod
+    def _get_base_command(self) -> list[str]:
+        pass
```
Comment (gchwier) [inline scripts/pylib/twister/twisterlib/harness.py:308]: add docstring instead of pass

File: scripts/pylib/twister/twisterlib/harness.py [hunk 3]
```diff
@@ -507,6 +487,116 @@ def _parse_report_file(self, report):
...
+            pytest_plugin_path = os.path.join(ZEPHYR_BASE, 'scripts', 'pylib',
+                                              'pytest-twister-harness',
+                                              'src')
+
+            command.extend(['--pythonpath', pytest_plugin_path])
+
+        command.append(os.path.normpath(os.path.join(self.source_dir, os.path.expanduser(os.path.expandvars(robot_test_path)))))
+
+
+class PytestHarnessException(TwisterLibraryHarnessException):
```
Comment (gchwier) [inline scripts/pylib/twister/twisterlib/harness.py:543]: not used now

File: scripts/pylib/twister/twisterlib/harness.py [hunk 4]
```diff
@@ -284,60 +286,36 @@ def handle(self, line):
...
 
-class PytestHarnessException(Exception):
-    """General exception for pytest."""
+class TwisterLibraryHarnessException(Exception):
+    """General exception for twister library harness usage."""
 
+class TwisterLibraryHarness(Harness, abc.ABC):
 
-class Pytest(Harness):
+    def _configure(self, instance: TestInstance, tool_name: str, report_filename: str, log_filename: str):
```
Comment (gchwier) [inline scripts/pylib/twister/twisterlib/harness.py:294]: why not `configure` (without underscore)? It is base method. Then `toolname`, `report_filename` and `log_filename` could be updated in `configure` methods in child classes

**Comment by MP-StefanKraus** [comment]
Thanks nashif, I fixed the CI by simulate_exclude 'renode' from the SIMS of pytest.

File: scripts/pylib/twister/twisterlib/harness.py [hunk 5]
```diff
@@ -507,6 +487,116 @@ def _parse_report_file(self, report):
...
+            command.extend([
+                '--loglevel', 'DEBUG',
+            ])
+
+        if not PYTEST_PLUGIN_INSTALLED:
+            pytest_plugin_path = os.path.join(ZEPHYR_BASE, 'scripts', 'pylib',
+                                              'pytest-twister-harness',
+                                              'src')
+
+            command.extend(['--pythonpath', pytest_plugin_path])
```
Comment (gchwier) [inline scripts/pylib/twister/twisterlib/harness.py:538]: probably do not need it after calling `_update_command_with_env_dependencies` (is called in `run_command` and adds twister-harness dir to PYTHONPATHS)

File: scripts/pylib/twister/twisterlib/runner.py [hunk 6]
```diff
@@ -1096,6 +1096,8 @@ def run(self):
             #
             if isinstance(harness, Pytest):
                 harness.pytest_run(instance.handler.get_test_timeout())
+            elif isinstance(harness, Robotframework):
+                harness.robot_run(instance.handler.get_test_timeout())
```
Comment (gchwier) [inline scripts/pylib/twister/twisterlib/runner.py:1100]: maybe better to check if it is `TwisterLibraryHarness` and call common method `run` (both `pytest_run` and `robot_run` are the same

File: scripts/pylib/twister/twisterlib/harness.py [hunk 7]
```diff
@@ -507,6 +487,116 @@ def _parse_report_file(self, report):
...
+        super(Robotframework, self)._configure(instance, "robot", 'robot_report.xml', 'robot.html')
+
+    def robot_run(self, timeout):
+        self._run(timeout)
+
+    def _get_base_command(self):
+        command = [
+            sys.executable,
+            "-m",
+            "twister_harness.robot_framework.robot_wrapper",
```
Comment (gchwier) [inline scripts/pylib/twister/twisterlib/harness.py:502]: Pytest plugin (and pytest also) is not used here, only a module, that is located in twister_harness dir. So maybe that module should be moved to outside of pytest-twister-harness plugin?

**Review by gchwier — CHANGES_REQUESTED** [review]
`robot_wrapper` only uses part of `pytest-twister-harness`. Therefore, it should not be a part of the plugin. Instead, it should be moved to a separate location (e.g. `zephyr/scripts/pylib/robot_framework`) ... and import required modules.
This solution causes high coupling between pytest plugin and robot framework, which should be taken into account in future development.

File: samples/subsys/testsuite/robotframework/shell/src/main.c [hunk 8]
```diff
@@ -0,0 +1,14 @@
+/*
```
Comment (hakehuang) [inline samples/subsys/testsuite/robotframework/shell/src/main.c:1]: why we can not reuse existing shell samples here?

**Review by hakehuang — COMMENTED** [review]
does this mean all robot cases need put into robotframework? this is not good in practices. Ideally, we expect robotframework and pytest or gtest with the same test code context, instead to move same test code around folders. Am I get it wrong?

**Comment by mateusz-holenko** [comment]
Hi @MP-StefanKraus, as mentioned in https://github.com/zephyrproject-rtos/zephyr/issues/67608#issuecomment-1893865296 we believe that instead of introducing a separate `robotframework` harness it could be beneficial to generalize the existing `robot` one and allow users to select the underlying robot runner.
As a result we would have a single mechanism that could target hardware and simulated targets (being it Renode or other simulators).

**Comment by MP-StefanKraus** [comment]
Hi,

thanks all for the comments and input on this PR!

As it shows, there are some general things we need to figure out first (re-using the existing one vs separate harness, robot naming in pytest-twister-harness, and others), which will be discussed in the RFC https://github.com/zephyrproject-rtos/zephyr/issues/67608;

I therefore set this PR on 'Draft' and will rework it after the plan is set. 

Thanks a lot so far,

**Pull request closed without merging**